### PR TITLE
Added new module gitlab_tag

### PR DESCRIPTION
**Pull request by dj-wasabi**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

gitlab_tag
##### ANSIBLE VERSION

```
ansible 2.1.2.0
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

This module will create a git tag in Gitlab on a project. The group and project should exists, otherwise this module will complain that either one of these does not exists. This module can only create or delete a tag, not update a tag.

### CODE REVIEW

**Comment by dj-wasabi** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/3283

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    project: my_first_project
+    name: v1.0
+    ref: master
+    description: My First Release
+    state: present
+
+- name: "Delete tag"
+  local_action:
+    module: gitlab_tag
+    server_url="http://gitlab.dj-wasabi.local"
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:110]: Examples should use HTTPS

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        default: "present"
+        choices: ["present", "absent"]
+
+'''
+
+EXAMPLES = '''
+- name: "Create tags"
+  local_action:
+    module: gitlab_tag
+    server_url="http://gitlab.dj-wasabi.local"
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:97]: Examples should use HTTPS

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+RETURN = '''# '''
+
+try:
+    import gitlab
+    HAS_GITLAB_PACKAGE = True
+except:
+    HAS_GITLAB_PACKAGE = False
+
+from ansible.module_utils.basic import *
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:128]: I believe importing all symbols from `ansible.modules_utils.basic` is now considered bad practice. You should import only what you need.

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+            validate_certs=dict(required=False, default=True, type='bool', aliases=['verify_ssl']),
+            login_user=dict(required=False, no_log=True),
+            login_password=dict(required=False, no_log=True),
+            login_token=dict(required=False, no_log=True),
+            group=dict(required=True),
+            project=dict(required=True),
+            name=dict(required=True),
+            ref=dict(required=False, default="master"),
+            description=dict(required=False),
+            state=dict(default="present", choices=["present", 'absent']),
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:201]: This list would be easier to grok if it was in alphabetical order.

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        use_credentials = False
+    else:
+        module.fail_json(msg="No login credentials are given. Use login_user with login_password, or login_token")
+
+    if login_token and login_user:
+        module.fail_json(msg="You can either use 'login_token' or 'login_user' and 'login_password'")
+
+    try:
+        if use_credentials:
+            git = gitlab.Gitlab(server_url, email=login_user, password=login_password, ssl_verify=verify_ssl)
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:235]: I think `gitlab` would be a better name for this variable than `git`.

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    login_user = module.params['login_user']
+    login_password = module.params['login_password']
+    login_token = module.params['login_token']
+    project = module.params['project']
+    group = module.params['group']
+    name = module.params['name']
+    ref = module.params['ref']
+    description = module.params['description']
+    state = module.params['state']
+    use_credentials = None
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:220]: A lot of this code is duplicated between this `gitlab_tag` module and the `gitlab_hook` module that you've also submitted: https://github.com/ansible/ansible/pull/19093

I think you should extract the common code into a class that goes into `ansible.module_utils`.

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    if not project:
+        module.fail_json(msg="The project or the group does not exists. We can not create a tag.")
+
+    project_tag = tag.existsTag(name=name)
+    if project_tag:
+        if state == "present":
+            module.exit_json(changed=False, result="Tag is already created.")
+        else:
+            if module.check_mode:
+                module.exit_json(changed=True, result="Tag should have been deleted.")
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:255]: Change `should` to `would`.

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+            if tag.deleteTag():
+                module.exit_json(changed=True, result="Tag is deleted.")
+            else:
+                module.exit_json(changed=False, result="Something went wrong deleting the tag.")
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py]: I would use `module.fail_json` here to properly indicate failure.

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+    else:
+        if state == "absent":
+            module.exit_json(changed=False, result="There is no tag available.")
+        else:
+            if module.check_mode:
+                module.exit_json(changed=True, result="Tag should have been created.")
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:263]: Change `should` to `would`.

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+            if tag.createTag(name=name, ref=ref, description=description):
+                module.exit_json(changed=True, result="Tag is created.")
+            else:
+                module.exit_json(changed=False, result="Something went wrong creating the tag.")
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:267]: As above, use `module.fail_json` to properly indicate failure.

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+class GitLabTag(object):
+    def __init__(self, module, git):
+        self._module = module
+        self._gitlab = git
+        self.projectObject = None
+        self.tagObject = None
+
+    def existsTag(self, name):
+        """Validate is tag exists"""
+        project = self.projectObject
```
Comment (djmattyg007) [inline lib/ansible/modules/source_control/gitlab_tag.py:141]: If something calls `existsTag` before `getProject` is called, `project` will hold `None`, and will then cause problems. I think there should be a `GitLabProject` class (which should also go into the `ansible.module_utils` namespace) which should be passed in to this class through the constructor.

**Review by djmattyg007 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/source_control/gitlab_tag.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+---
+module: gitlab_tag
+short_description: Creates/deletes tags on Gitlab projects
```
Comment (adrianmoisey) [inline lib/ansible/modules/source_control/gitlab_tag.py:22]: This is a nitpick, I'm not 100% sure if it's right, but it looks like GitLab like to go by "GitLab" and not "Gitlab"

**Pull request closed without merging**